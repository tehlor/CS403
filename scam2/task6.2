(println (powerSet '(y w e b u r)))
