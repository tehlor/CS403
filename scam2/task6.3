(println (powerSet '(z k t v u)))
