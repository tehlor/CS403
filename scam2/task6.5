(println (powerSet '(u r i p e w b)))
