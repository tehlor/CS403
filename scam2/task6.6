(println (powerSet '(l w k e)))
