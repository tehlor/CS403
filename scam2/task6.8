(println (powerSet '(u h g q a p t y l)))
