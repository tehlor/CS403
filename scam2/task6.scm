(define (main)
    (setNilDisplay 'nil)
    (setPort (open (getElement ScamArgs 1) 'read))
    (readArgs)
)

(define (readArgs)
    (define env this)
    (define (iter expr)
        (if (not (eof?))
            (begin (eval expr env) 
                (iter (readExpr))
            )
        )
    )
    (iter (readExpr))
)
;; let psetRest be all obtained subsets (ps (cdr set)) so we dont have to call it twice below 
;; append the list of all current subsets to the new subsets obtained by adding (car set) to all current subsets
;; cons car to all subsets of pset rest to obtain the new subsets
(define (F e T)
  (map
    (lambda (X)
      (cons e X)
      )
    T
    )
  )

(define (powerSet S)
  (cond
    ((null? S) '(()) )
    (else 
      (let ( (T (powerSet (cdr S))) (e (car S)) )
        (merge
          T
          (F e T)
          )
        )
      )
    )
  )

(define (merge a b)
  (define (help f s)
    (cond 
      ((null? f) s)
      ((null? s) f)
      (else
            (if (< (length (car f)) (length (car s)))
                (cons (car f) (help (cdr f) s))
                (cons (car s) (help f (cdr s)))
            )
        )
      )
    )
  (help a b)
)

