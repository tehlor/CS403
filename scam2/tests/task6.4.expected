(nil (c) (g) (z) (w) (x) (c g) (c z) (c w) (c x) (g z) (g w) (g x) (z w) (z x) (w x) (c g z) (c g w) (c g x) (c z w) (c z x) (c w x) (g z w) (g z x) (g w x) (z w x) (c g z w) (c g z x) (c g w x) (c z w x) (g z w x) (c g z w x))
